dev/run

```diff
@@ -264,6 +278,10 @@ def setup_context(opts, args):
         "auto_ports": opts.auto_ports,
         "locald_configs": opts.locald_configs,
         "with_nouveau": opts.with_nouveau,
+        "enable_tls": opts.enable_tls,
+        "tcp_node": opts.tcp_node,
+        "hostname": socket.gethostname() if opts.enable_tls else "127.0.0.1",
+        "bind_address": "0.0.0.0" if opts.enable_tls else "127.0.0.1",
```

Comment:
why is this necessary? Ideally the dev nodes remain bound only to 127.0.0.1. Is something in openssl forcing this?

Reply:
Do you need to override server_name_indication? I thought the CN once set to the erlang node name would be the right value. did that not work? (`"/CN=node1@127.0.0.1"`, I mean.)

Reply:
np, `node1@127.0.0.1` does not work. CN needs to be equal to the `hostname`.
e.g.: If the node name is `ssl_test@localhost`, CN needs to equal `localhost`.

I've tried different CNs, such as 127.0.0.1, {127,0,0,1}, 127.0.1.1, localhost, localhost.local, `$hostname -s/-f`, etc. If the node name contains 127.0.0.1, they just return `hostname_check_failed` in OTP 25.

Reply:
it's more complicated than that. reading `inet_tls_dist.erl` makes it clear that they expect the host name part in the SAN field and the node name in a utf8string as a commonName under a directoryName. I have not yet persuaded openssl to generate a certificate that `inet_tls_dist:cert_nodes/1` is happy with. I think the author(s) of `inet_tls_dist` has made a cryptic choice;

```
%% Look in Extensions, in all subjectAltName:s
%% to find node names in this certificate.
%% Host names are picked up as a subjectAltName containing
%% a dNSName, and the first subjectAltName containing
%% a commonName is the node name.
%%
```

The nearest I can get so far is;

```
 [{'Extension',{2,5,29,17},
                                                     false,
                                                     [{dNSName,"127.0.0.1"},
                                                      {directoryName,{rdnSequence,[[{'AttributeTypeAndValue',{2,5,
                                                                                                              4,3},
                                                                                                             <<12,7,99,111,117,99,104,100,98>>}]]}}]},

```

`cert_nodes` finds the hostname but because the `<<12,7,99,111,117,99,104,100,98>>` is not a `utf8String` tuple it doesn't find the node name.

This is the (bizarre) way I have managed to add the dirName to the certificate. all other methods have failed (like the invitingly simple `-addext`).

```
printf "[SAN]\nsubjectAltName=DNS:127.0.0.1,dirName:dir_ext\n[dir_ext]\nCN=couchdb\n" > ext.cnf
openssl x509 -req -days 3650 -set_serial 01 -in node-req.pem -out node-cert.pem -CA ca-cert.pem -CAkey ca-key.pem -extfile ext.cnf -extensions SAN 
```

Thus far I think the openssl cli is mediocre and you can't actually do everything there that openssl is capable of. It seems you _must_ use openssl conf files to do anything interesting.

This has, so far, been frustrating and illuminating.

Reply:
https://github.com/rlipscombe/elixir-certs/blob/main/certs.exs looks helpful...

Reply:
Thank you, Bob.

Reply:
Last bit of the mystery is clearing. There is a variant or extended certificate format specific to OTP. If I use that then the dirName is encoded correctly. openssl is obviously unaware of the OTP-specific `OTPCertificate` format, so it cannot be used to generate certificates for inet_tls_dist that would check both node and host name (for sample erlang node name `node@host`).

```
X509.RDNSequence.new("/CN=Bob")
{:rdnSequence,
 [[{:AttributeTypeAndValue, {2, 5, 4, 3}, <<12, 3, 66, 111, 98>>}]]}

X509.RDNSequence.new("/CN=Bob", :otp)
{:rdnSequence, [[{:AttributeTypeAndValue, {2, 5, 4, 3}, {:utf8String, "Bob"}}]]}
```

The rabbit hole is deep and full of terrors.

Reply:
This diff to https://github.com/rlipscombe/elixir-certs allows me to generate a correct certificate.

```
diff --git a/certs.exs b/certs.exs
index 86fb2d2..e52a436 100644
--- a/certs.exs
+++ b/certs.exs
@@ -78,7 +78,8 @@ defmodule Certs do
         ext_key_usage: ext_key_usage([:serverAuth, :clientAuth]),
         subject_key_identifier: true,
         authority_key_identifier: true,
-        subject_alt_name: subject_alt_name([commonName])
+        subject_alt_name: subject_alt_name([commonName]),
+        subject_alt_name: subject_alt_name([{:directoryName, X509.RDNSequence.new(subject, :otp)}])
       ]
     }
   end
```

yields a certificate that ```
5> inet_tls_dist:cert_nodes(OTPCert).
["server@server"]
```

Obviously the values are wrong, the first `server` should be `couchdb` and the second server should be `127.0.0.1` but those are just strings.

Reply:
what we'd actually need;

```
                                        {'Extension',{2,5,29,17},false,[{dNSName,"127.0.0.1"}]},
                                        {'Extension',{2,5,29,17},
                                                     false,
                                                     [{directoryName,{rdnSequence,[[{'AttributeTypeAndValue',{2,
                                                                                                              5,4,3},
                                                                                                             {utf8String,<<"node1">>}}]]}}]}]},
                  {'SignatureAlgorithm',{1,2,840,10045,4,3,2},asn1_NOVALUE},
                  <<48,69,2,32,105,178,12,49,33,238,187,152,243,237,155,
                    128,86,97,241,117,150,223,147,14,210,204,126,69,46,255,
                    22,12,165,28,214,174,2,33,0,133,247,16,211,97,56,77,
                    164,21,139,198,41,184,243,254,91,25,64,39,242,67,179,
                    228,255,131,202,213,27,38,91,54,36>>}
ok
5>
5> inet_tls_dist:cert_nodes(OTPCert).
["node1@127.0.0.1"]
```

Reply:
forked the elixir-certs repo and added options to generate these things: https://github.com/rnewson/elixir-certs/

Reply:
a remsh using inet_tls_dist but without presenting a client certificate is rejected as expected;

```
erl -boot ./start_ssl -name remsh$$@127.0.0.1 -hidden -proto_dist inet_tls  -remsh node1@127.0.0.1
Erlang/OTP 24 [erts-12.3.2.9] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit]

=WARNING REPORT==== 25-May-2023::13:09:35.881585 ===
Description: "Authenticity is not established by certificate path validation"
     Reason: "Option {verify, verify_peer} and cacertfile/cacerts is missing"

*** ERROR: Shell process terminated! (^G to start new job) ***
=NOTICE REPORT==== 25-May-2023::13:09:36.033411 ===
TLS client: In state cipher received SERVER ALERT: Fatal - Handshake Failure

(node1@127.0.0.1)1> =NOTICE REPORT==== 25-May-2023::13:09:36.023069 ===
TLS server: In state certify at tls_dtls_connection.erl:315 generated SERVER ALERT: Fatal - Handshake Failure
 - no_client_certificate_provided
```

Reply:
Thanks for all the info, especially for generating the correct certificates to pass validation!